specification/flag-evaluation/hooks.md

```diff
@@ -2,4 +2,126 @@
 
 ## Overview
 
-// TODO: this is a stub.
+Hooks are a mechanism whereby application developers can add arbitrary behavior
+to flag evaluation. They operate similarly to middleware in many web frameworks.
+
+### Definitions
+
+**Stage**: An explicit portion of the flag evaluation lifecycle. e.g. `before` being "before the evaluation is run.
+**Invocation**: A single call to evaluate a flag. `client.getBooleanValue(..)` is an invocation.
+**API**: The global API singleton.
+
+### Hook context
+
+Hook context exists to provide hooks with information about the invocation.
+
+> Hook context **MUST** provide: the flag key, evaluation context, default value and a list of  executed hooks by stage.
+
+> Condition: You **MUST** provide `flag type` if the language type system differentiates between strings, numbers, booleans, and structures.
+
+> Hook context **SHOULD** provide: provider, client
+
+> flag key, flag type, default value properties **MUST** be immutable.
```

Comment:
maybe add something here like "if the language does not support immutability, the hook MUST NOT modify these properties"